### ISSUE TITLE
The web browser opens before ui.run()

### ISSUE THREAD

**Issue by JS-Aibel** [zauberzeug/nicegui#412]
I do not know if this issue is related to https://github.com/zauberzeug/nicegui/issues/159

A browser tab is opened before the Python script runs the ui.run() command. This result in a "can't reach this page" in the browser. When the terminal is showing “NiceGUI ready to go on […]”, the user needs to click “refresh” in the browser tab to show the application.

The problem seems to occur  when there is a lot of work to be done before the script runs the command ui.run().
Tested on Edge and Fire fox.

**Example code:**
```py
from nicegui import ui

ui.label('some label')

for i in range(150000):
    print(i)

ui.run()
```

**Comment by rodja** [comment]
Looks like the same issue than #159. But on Mac it's not reproducable with Chrome or Firefox. Both open early but then retry connecting until the port is finally opened by Uvicorn. Regardless it would be great to do

https://github.com/zauberzeug/nicegui/blob/c5ce65a2141a477d630acb2814cd2afb29bc1148/nicegui/run.py#L68-L69

after we started the server and present the "NiceGUI ready to go" message.

**Comment by JS-Aibel** [comment]
Thanks for the reply. 
When I set the for loop down to 100000, I get the same result as you - the browser tab opens early, but then it retries connecting. So then there is no need to click refresh in the browser.

For a more independent benchmark, I replaced the for loop in my test with `time.sleep()` and I got the following results.

- `time.sleep(10)` – Browser tab opens early, but connects automatically after terminal shows “NiceGUI ready to go on […]”

- `time.sleep(15)` – Browser tab opens early, and the user have to click “refresh” after “NiceGUI ready to go on […]”.

**Comment by rodja** [comment]
Even with `time.sleep(15)` I do not need to click refresh. But I get the point.

**Comment by rodja** [comment]
@JS-Aibel could you try out the pull request #435 and verify that it fixes the issues?

**Comment by JS-Aibel** [comment]
> @JS-Aibel could you try out the pull request #435 and verify that it fixes the issues?

Yes, it fixed the issue.

**Comment by rodja** [comment]
As @falkoschindler wrote in https://github.com/zauberzeug/nicegui/pull/435#issuecomment-1446121074, the proposed solution has other issues. For example opening the browser every time code is modified (with auto-reload enabled)...

**Comment by falkoschindler** [comment]
I tried reproducing it with Chrome, Safari, and Firefox on Mac OS, but without success.

As far as I understand, the procedure is as follows:

- The main script is evaluated. The slow loop (or `time.sleep`) is executed a first time.
- `ui.run` opens the browser.
- `ui.run` starts the uvicorn server. This is where the main process blocks.
- A secondary process (in case of `reload=True`) evaluates the main script. The loop is executed a second time.
- In the meantime the browser tab might give up early and the user needs to manually reload the page.

The question is: How can the main process postpone opening the browser until the secondary process is ready? How could it even know _how long_ it needs to wait?

Maybe we need to conclude that costly setup code needs to run asynchronously to not block the process.

**Comment by tobb10001** [comment]
I might have found something useful.

> How could it even know how long it needs to wait?

One could use a TCP socket to find out if the port that `ui.run()` listens on already accepts TCP connections. Similar to a really rudimentary portscanner would do it.

I've got this to work in an experiment; I could prepare a PR if you think this idea is any good.

**Comment by rodja** [comment]
Oh that sounds super cool. Please create the PR. We are eager to review/merge if it works reliable.

**Comment by rodja** [comment]
Maybe this is also related to #607 where we want to set the webviews content only after the server has started?
